### PR TITLE
updates gorm user dao

### PR DESCRIPTION
**Pull request by horlakz**
- **chore(deps): add gorm and jinzhu libraries to go.mod**
- **feat(orms/gorm): implement user CRUD operations and payload structures**

### CODE REVIEW

File: orms/gorm/user.go [hunk 1]
```diff
@@ -1,58 +1,116 @@
...
-type UserCreatePayload struct{}
-
-type UserUpdatePayload struct{}
+type UserUpdatePayload struct {
+	Id        *string `json:"id"`
+	FirstName *string `json:"firstName"`
+	LastName  *string `json:"lastName"`
+	Email     *string `json:"email"`
+	Username  *string `json:"username"`
+	Password  *string `json:"password"`
```
Comment (struckchure) [inline orms/gorm/user.go:29]: needs to be hidden on the json layer

File: orms/gorm/user.go [hunk 2]
```diff
@@ -1,58 +1,116 @@
...
 
+import "gorm.io/gorm"
+
 type User struct {
 	Id        string `json:"id"`
+	FirstName string `json:"firstName" gorm:"column:first_name"`
+	LastName  string `json:"lastName" gorm:"column:last_name"`
+	Email     string `json:"email"`
+	Username  string `json:"username"`
+	Password  string `json:"password"`
```
Comment (struckchure) [inline orms/gorm/user.go]: should be hidden from the json output

Comment (horlakz) [reply orms/gorm/user.go]: the password ba???
normally I usually have a userdto in my service to transform response

File: orms/gorm/user.go [hunk 3]
```diff
@@ -1,58 +1,116 @@
...
 	LastName  string `json:"lastName"`
 	Email     string `json:"email"`
 	Username  string `json:"username"`
 	Password  string `json:"password"`
 }
 
-type UserCreatePayload struct{}
-
-type UserUpdatePayload struct{}
+type UserUpdatePayload struct {
```
Comment (struckchure) [inline orms/gorm/user.go]: check the update pattern in the prisma dao

File: orms/gorm/user.go [hunk 4]
```diff
@@ -1,58 +1,116 @@
...
+	Username  string `json:"username"`
+	Password  string `json:"-"`
+	CreatedAt string `json:"created_at" gorm:"column:created_at"`
+}
+
+type UserCreatePayload struct {
 	FirstName string `json:"firstName"`
 	LastName  string `json:"lastName"`
 	Email     string `json:"email"`
 	Username  string `json:"username"`
```
Comment (struckchure) [inline orms/gorm/user.go]: remove username field too

File: orms/gorm/user.go [hunk 5]
```diff
@@ -1,58 +1,116 @@
...
+	FirstName string `json:"firstName" gorm:"column:first_name"`
+	LastName  string `json:"lastName" gorm:"column:last_name"`
+	Email     string `json:"email"`
+	Username  string `json:"username"`
+	Password  string `json:"-"`
+	CreatedAt string `json:"created_at" gorm:"column:created_at"`
+}
+
+type UserCreatePayload struct {
 	FirstName string `json:"firstName"`
```
Comment (struckchure) [inline orms/gorm/user.go]: first and last name should be of type `*string`

File: orms/gorm/user.go [hunk 6]
```diff
@@ -1,58 +1,116 @@
 package gorm
 
+import "gorm.io/gorm"
+
 type User struct {
 	Id        string `json:"id"`
+	FirstName string `json:"firstName" gorm:"column:first_name"`
```
Comment (struckchure) [inline orms/gorm/user.go]: is there a way to specify nullable fields in gorm?

File: orms/gorm/user.go [hunk 7]
```diff
@@ -1,58 +1,116 @@
...
+	LastName  string `json:"lastName" gorm:"column:last_name"`
+	Email     string `json:"email"`
```
Comment (struckchure) [inline orms/gorm/user.go]: Make this unique too

File: orms/gorm/user.go [hunk 8]
```diff
@@ -1,58 +1,116 @@
...
-	Get(string) (*User, error)
-	GetByEmail(string) (*User, error)
-	GetByUsername(string) (*User, error)
-	Create(UserCreatePayload) (*User, error)
-	Update(UserUpdatePayload) (*User, error)
+	Get(string) (User, error)
+	GetByEmail(string) (User, error)
+	GetByUsername(string) (User, error)
+	Create(UserCreatePayload) (User, error)
+	Update(UserUpdatePayload) (User, error)
```
Comment (struckchure) [inline orms/gorm/user.go]: this should be `Update(string, UserUpdatePayload) (User, error)`

**Review by struckchure — CHANGES_REQUESTED** [review]

File: orms/gorm/user.go (hunk 6 above)
Comment (horlakz) [reply orms/gorm/user.go]: yes sure

File: orms/gorm/user.go (hunk 5 above)
Comment (horlakz) [reply orms/gorm/user.go]: optional

**Pull request merged**